Index: Classes/Middleware/RedirectionMiddleware.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Classes/Middleware/RedirectionMiddleware.php b/Classes/Middleware/RedirectionMiddleware.php
--- a/Classes/Middleware/RedirectionMiddleware.php	
+++ b/Classes/Middleware/RedirectionMiddleware.php	(date 1645460312634)
@@ -44,6 +44,17 @@
          */
         $cookieName = 'site-language-preference';
 
+		// stop if somewhere in our rootline tx_sitelanguageredirection_stop has been set
+		$pageArguments = $request->getAttribute('routing');
+		$rootlineUtility = GeneralUtility::makeInstance(RootlineUtility::class, $pageArguments->getPageId());
+		/** @var RootlineUtility $rootlineUtility */
+		$pageRootline = $rootlineUtility->get();
+		foreach ($pageRootline as $pageItem) {
+			if (isset($pageItem['tx_sitelanguageredirection_stop']) && $pageItem['tx_sitelanguageredirection_stop'] === 1) {
+				return $handler->handle($request);
+			}
+		}
+
         /**
          * Do not redirect search engine bots.
          */
Index: Resources/Private/Language/locallang_db.xlf
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Resources/Private/Language/locallang_db.xlf b/Resources/Private/Language/locallang_db.xlf
new file mode 100644
--- /dev/null	(date 1645460311134)
+++ b/Resources/Private/Language/locallang_db.xlf	(date 1645460311134)
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="utf-8" standalone="yes" ?>
+<xliff version="1.0">
+	<file source-language="en" datatype="plaintext" original="messages" date="2019-08-09T08:33:58Z">
+		<header>
+			<type>database</type>
+			<description>Language labels for database tables/fields belonging to extension 'site-language-redirection'</description>
+			<generator>LFEditor</generator>
+			<authorName>Matthia Adrowski</authorName>
+			<authorEmail>matthias.adrowski@sgalinski.de</authorEmail>
+		</header>
+		<body>
+			<trans-unit id="pages.tx_sitelanguageredirection_stop.label" xml:space="preserve">
+				<source>Stop Site Language Redirects</source>
+			</trans-unit>
+		</body>
+	</file>
+</xliff>
Index: Resources/Private/Language/de.locallang_db.xlf
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Resources/Private/Language/de.locallang_db.xlf b/Resources/Private/Language/de.locallang_db.xlf
new file mode 100644
--- /dev/null	(date 1645460311114)
+++ b/Resources/Private/Language/de.locallang_db.xlf	(date 1645460311114)
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8" standalone="yes" ?>
+<xliff version="1.0">
+	<file source-language="en" datatype="plaintext" original="messages" date="2019-08-09T08:33:58Z">
+		<header>
+			<type>database</type>
+			<description>Language labels for database tables/fields belonging to extension 'site-language-redirection'</description>
+			<generator>LFEditor</generator>
+			<authorName>Matthia Adrowski</authorName>
+			<authorEmail>matthias.adrowski@sgalinski.de</authorEmail>
+		</header>
+		<body>
+			<trans-unit id="pages.tx_sitelanguageredirection_stop.label" xml:space="preserve">
+				<source>Stop Site Language Redirects</source>
+				<target>Stop Site Language Redirects</target>
+			</trans-unit>
+		</body>
+	</file>
+</xliff>
Index: Configuration/TCA/Overrides/pages.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Configuration/TCA/Overrides/pages.php b/Configuration/TCA/Overrides/pages.php
new file mode 100644
--- /dev/null	(date 1645460299182)
+++ b/Configuration/TCA/Overrides/pages.php	(date 1645460299182)
@@ -0,0 +1,53 @@
+<?php
+/**
+ *
+ * Copyright notice
+ *
+ * (c) sgalinski Internet Services (https://www.sgalinski.de)
+ *
+ * All rights reserved
+ *
+ * This script is part of the TYPO3 project. The TYPO3 project is
+ * free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The GNU General Public License can be found at
+ * http://www.gnu.org/copyleft/gpl.html.
+ *
+ * This script is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * This copyright notice MUST APPEAR in all copies of the script!
+ */
+
+\TYPO3\CMS\Core\Utility\ExtensionManagementUtility::addTCAcolumns(
+	'pages',
+	[
+		'tx_sitelanguageredirection_stop' => [
+			'exclude' => 1,
+			'l10n_mode' => 'exclude',
+			'label' => 'LLL:EXT:site_language_redirection/Resources/Private/Language/locallang_db.xlf:pages.tx_sitelanguageredirection_stop.label',
+			'config' => [
+				'type' => 'check',
+				'renderType' => 'checkboxToggle',
+				'items' => [
+					[
+						0 => '',
+						1 => '',
+					]
+				],
+			]
+		],
+	]
+);
+
+\TYPO3\CMS\Core\Utility\ExtensionManagementUtility::addToAllTCAtypes(
+	'pages',
+	'tx_sitelanguageredirection_stop',
+	1,
+	'after:php_tree_stop'
+);
Index: ext_tables.sql
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/ext_tables.sql b/ext_tables.sql
new file mode 100644
--- /dev/null	(date 1645460335986)
+++ b/ext_tables.sql	(date 1645460335986)
@@ -0,0 +1,3 @@
+CREATE TABLE pages (
+	tx_sitelanguageredirection_stop tinyint(4) unsigned DEFAULT 0 NOT NULL,
+);
Index: ext_localconf.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/ext_localconf.php b/ext_localconf.php
new file mode 100644
--- /dev/null	(date 1645460333590)
+++ b/ext_localconf.php	(date 1645460333590)
@@ -0,0 +1,7 @@
+<?php
+
+call_user_func(
+	static function () {
+		$GLOBALS['TYPO3_CONF_VARS']['FE']['addRootLineFields'] .= ', tx_sitelanguageredirection_stop';
+	}
+);
