diff --git a/Classes/Middleware/RedirectionMiddleware.php b/Classes/Middleware/RedirectionMiddleware.php
index 7662bda..0ffd0a7 100644
--- a/Classes/Middleware/RedirectionMiddleware.php
+++ b/Classes/Middleware/RedirectionMiddleware.php
@@ -186,6 +186,10 @@ class RedirectionMiddleware implements MiddlewareInterface
 
         $parameters['_language'] = $matchingSiteLanguage;
 
+		$iIdLang = (int)$matchingSiteLanguage->getLanguageId();
+		if (!$this->checkPageAvailableInLanguage($pageArguments->getPageId(), $iIdLang)) {
+			return null;
+		}
         $uri = $site->getRouter()->generateUri(
             $pageArguments->getPageId(),
             $parameters
@@ -384,4 +388,40 @@ class RedirectionMiddleware implements MiddlewareInterface
 
         return is_string($userAgent) && preg_match($this->botPattern, $userAgent);
     }
+
+	/**
+	 * @param int $iPageId
+	 * @param int $iIdLang
+	 * @return bool
+	 */
+	protected function checkPageAvailableInLanguage(int $iPageId, int $iIdLang): bool
+	{
+		if ($iIdLang > 0) {
+			$table = 'pages';
+			$queryBuilder = GeneralUtility::makeInstance(\TYPO3\CMS\Core\Database\ConnectionPool::class)
+				->getQueryBuilderForTable($table);
+			$data = $queryBuilder
+				->select('uid')
+				->from($table)
+				->where(
+					$queryBuilder->expr()->eq('l10n_parent', $iPageId)
+				)
+				->andWhere(
+					$queryBuilder->expr()->eq('sys_language_uid', $iIdLang)
+				)
+				->andWhere(
+					$queryBuilder->expr()->eq('deleted', 0)
+				)
+				->andWhere(
+					$queryBuilder->expr()->eq('hidden', 0)
+				)
+				->setMaxResults(1)
+				->execute()
+				->fetchAll();
+			if (count($data) == 0) {
+				return false;
+			}
+		}
+		return true;
+	}
 }
